internal/testintegration/setup.go

```diff
@@ -110,85 +107,143 @@ func EnvTestSetup() (func(), error) {
-
-	Client, err = client.New(cfg, client.Options{Scheme: s})
+	// Initialize rest client configuration
+	th.s = scheme.Scheme
+	controller.InitScheme(th.s)
+	cl, err := client.New(cfg, client.Options{Scheme: th.s})
 	if err != nil {
-		return teardownFunc, fmt.Errorf("unable to start kuberenetes envtest %v", err)
-	}
-	if Client == nil {
-		return teardownFunc, fmt.Errorf("unable to start kuberenetes envtest %v", err)
+		return th, fmt.Errorf("unable to to create client %v", err)
 	}
+	th.Client = cl
 
-	mgrCancelFunc, err := startManager(ctx, cfg, s, workload.DefaultProxyImage)
+	// Start the controller-runtime manager
+	err = th.StartManager(workload.DefaultProxyImage)
 	if err != nil {
-		return teardownFunc, fmt.Errorf("unable to start kuberenetes envtest %v", err)
+		return th, fmt.Errorf("unable to start kuberenetes envtest %v", err)
 	}
 
-	RestartManager = func(defaultProxyImage string) error {
-		mgrCancelFunc()
-		mgrCancelFunc, err = startManager(ctx, cfg, s, defaultProxyImage)
-		return err
+	return th, nil
+}
+
+// EnvTestHarness enables integration tests to control the lifecycle of the
+// operator's controller-manager.
+type EnvTestHarness struct {
+
+	// Mgr is the manager
+	Mgr ctrl.Manager
+
+	// Client is the kubernetes client.
+	Client client.Client
+
+	// The actual EnvTest environment
+	testEnv *envtest.Environment
+
+	// testEnvCancel is the context cancel function for the testEnv
+	testEnvCancel context.CancelFunc
+
+	// cancel is the context cancel function for the manager
+	cancel context.CancelFunc
+
+	// stopped channel is closed when the manager actually stops
+	stopped chan int
+
+	// ctx is the manager's context
+	ctx context.Context
+
+	// cfg is the client configuration from envtest
+	cfg *rest.Config
+
+	// s is the client scheme
+	s          *runtime.Scheme
+	testEnvCtx context.Context
+}
+
+// Teardown closes the TestEnv environment at the end of the testcase.
+func (h *EnvTestHarness) Teardown() {
+	if h.testEnvCancel != nil {
+		h.testEnvCancel()
+	}
+	if h.testEnv != nil {
+		err := h.testEnv.Stop()
+		if err != nil {
+			Log.Error(err, "unable to stop envtest environment %v")
+		}
 	}
+}
 
-	return teardownFunc, nil
+// StopManager stops the controller manager and waits for it to exit, returning an
+// error if the controller manager does not stop within 1 minute.
+func (h *EnvTestHarness) StopManager() error {
+	if h.cancel != nil {
+		h.cancel()
+	}
+	h.cancel = nil
+	select {
+	case <-h.stopped:
+		return nil
+	case <-time.After(1 * time.Minute):
+		return errors.New("manager did not stop after 1 minute")
+	}
 }
 
-func startManager(ctx context.Context, cfg *rest.Config, s *runtime.Scheme, proxyImage string) (context.CancelFunc, error) {
-	ctx, managerCancelFunc := context.WithCancel(ctx)
+// StartManager starts up the manager, configuring it with the proxyImage.
+func (h *EnvTestHarness) StartManager(proxyImage string) error {
+	h.ctx, h.cancel = context.WithCancel(h.testEnvCtx)
```

Comment:
We don't need to assign into a struct field here for ctx. Instead if we used a local variable, we could still save reference to h.cancel.

Reply:
Fixed.